assets/js/base/hocs/with-reviews.js

```diff
@@ -0,0 +1,150 @@
+		}
+
+		componentDidMount() {
+			this.getReviews();
+		}
+
+		getDefaultArgs() {
+			const { attributes } = this.props;
+			const { overwrittenArgs } = this.state;
+			const { order, orderby } = getOrderArgs( attributes.orderby );
+			const { productId, reviewsOnPageLoad } = attributes;
+
+			return {
+				order,
+				orderby,
+				per_page: reviewsOnPageLoad,
+				product_id: productId,
+				...overwrittenArgs,
+			};
+		}
+
+		getReviews() {
+			const { reviewsOnPageLoad } = this.props.attributes;
+			const args = this.getDefaultArgs();
+
+			this.updateListOfReviews( args, reviewsOnPageLoad );
+		}
+
+		onChangeArgs( newArgs ) {
+			const { reviewsOnPageLoad } = this.props.attributes;
+			const args = {
+				...this.getDefaultArgs(),
+				...newArgs,
+			};
+
+			this.setState( {
+				overwrittenArgs: newArgs,
+			} );
+
+			this.updateListOfReviews( args, reviewsOnPageLoad ).then( () => {
+				speak( __( 'Reviews order updated.', 'woo-gutenberg-products-block' ) );
+			} );
+		}
+
+		appendReviews() {
+			const { reviewsOnLoadMore } = this.props.attributes;
+			const oldReviews = this.state.reviews;
+			const args = {
+				...this.getDefaultArgs(),
+				offset: oldReviews.length,
+				per_page: reviewsOnLoadMore,
+			};
+
+			this.updateListOfReviews( args, reviewsOnLoadMore, oldReviews ).then( ( { newReviews } ) => {
+				speak(
+					sprintf(
+						_n(
+							'%d review loaded.',
+							'%d reviews loaded.',
+							newReviews.length,
+							'woo-gutenberg-products-block'
+						),
+						newReviews.length,
+					)
+				);
+			} );
+		}
+
+		updateListOfReviews( args, reviewsOnLoad, oldReviews = [] ) {
+			const { totalReviews } = this.state;
+			const reviewsToLoad = Math.min( totalReviews - oldReviews.length, reviewsOnLoad );
+
+			this.setState( {
+				loading: true,
+				reviews: oldReviews.concat( Array( reviewsToLoad ).fill( {} ) ),
+			} );
+
+			return getReviews( args ).then( ( { reviews: newReviews, totalReviews: newTotalReviews } ) => {
+				this.setState( {
+					reviews: oldReviews.filter( ( review ) => Object.keys( review ).length ).concat( newReviews ),
+					totalReviews: newTotalReviews,
+					loading: false,
+					error: null,
+				} );
+
+				return { newReviews };
+			} ).catch( this.setError );
+		}
+
+		setError( apiError ) {
+			const error = typeof apiError === 'object' && apiError.hasOwnProperty( 'message' ) ? {
+				apiMessage: apiError.message,
+			} : {
+				apiMessage: null,
+			};
+
+			this.setState( { reviews: [], loading: false, error } );
+			speak(
+				__( 'There was an error loading the reviews.', 'woo-gutenberg-products-block' )
+			);
+		}
+
+		render() {
+			const { error, loading, reviews, totalReviews } = this.state;
+
+			return <OriginalComponent
+				{ ...this.props }
+				error={ error }
+				getReviews={ this.getReviews }
+				appendReviews={ this.appendReviews }
+				onChangeArgs={ this.onChangeArgs }
+				isLoading={ loading }
+				reviews={ reviews }
+				totalReviews={ totalReviews }
```

Comment:
The only thing that should be passed along here are the `reviews`, `totalReviews`, `error` and `isLoading` (along with any props you want to pass through).

I would also have the `LoadMoreButton` be something injected in the wrapped component via the `withReviews` hoc (overrideable via render prop).

Reply:
I think this is very similar to the discussion we had with the `withProduct()` HOC and the _Retry_ button, right?

If we move the `<LoadMoreButton>` from the block to the HOC, shouldn't we do the same with the `<ReviewOrderSelect>`, which triggers `onChangeOrderby` → `onChangeArgs` → `getReviews`? But that sounds like we would be moving presentational logic to the HOC. What if another component wanted to use `withReviews()` but instead of having a load more button it would want to use infinite scrolling? And what if instead of a select to change the order it wanted to have buttons? That would make our HOC unusable for those cases.